File: backend/java/src/main/java/com/michaelyi/personalwebsite/post/PostUtil.java
package com.michaelyi.personalwebsite.post;

import com.michaelyi.personalwebsite.util.StringUtil;
import org.springframework.web.multipart.MultipartFile;

import java.util.Date;

public class PostUtil {
        public static final int OPENING_H1_TAG_LENGTH = "<h1>".length();
        public static final int CLOSING_H1_TAG_LENGTH = "</h1>".length();

        public static Post constructPost(String text) {
                if (!StringUtil.isStringValid(text)) {
                        throw new IllegalArgumentException("Text is invalid");
                }

                int openingH1TagIndex = text.indexOf("<h1>");
                int closingH1TagIndex = text.indexOf("</h1>");

                if (openingH1TagIndex == -1 || closingH1TagIndex == -1) {
                        throw new IllegalArgumentException("Title cannot be blank");
                }

                String title = text.substring(OPENING_H1_TAG_LENGTH, closingH1TagIndex);
                String content = text.substring(
                        closingH1TagIndex + CLOSING_H1_TAG_LENGTH);

                if (!StringUtil.isStringValid(title)) {
                        throw new IllegalArgumentException("Title cannot be blank");
                }

                if (!StringUtil.isStringValid(content)) {
                        throw new IllegalArgumentException("Content cannot be blank");
                }

                String id = convertTitleToId(title);
                Date date = new Date();
                title = removeHtmlTags(title);

                return new Post(id, date, title, null, content);
        }

        private static String convertTitleToId(String title) {
                // replace spaces with hyphens & remove all non-alphabetic characters
                String id = title.toLowerCase()
                        .replace(" ", "-")
                        .replaceAll("[^a-z\\-]", "");

                // remove trailing hyphens
                if (id.charAt(id.length() - 1) == '-') {
                        id = id.substring(0, id.length() - 1);
                }

                return id;
        }

        private static String removeHtmlTags(String s) {
                return s.replaceAll("<[^>]*>", "");
        }

        public static byte[] getImage(MultipartFile image) {
                if (!isImageValid(image)) {
                        throw new IllegalArgumentException("Image is invalid");
                }

                byte[] res;

                try {
                        res = image.getBytes();
                } catch (Exception e) {
                        res = null;
                }

                return res;
        }

        private static boolean isImageValid(MultipartFile image) {
                return image != null
                        && !image.isEmpty()
                        && image.getSize() != 0
                        && StringUtil.isStringValid(
                        image.getOriginalFilename())
                        && image
                        .getOriginalFilename()
                        .matches(".*\\.(jpg|jpeg|png|webp)")
                        && image
                        .getContentType()
                        .matches("image/(jpeg|png|webp)");
        }
}

File: backend/go/post/post_util.go
package post

import (
	"errors"
	"regexp"
	"strings"
	"time"

	"github.com/michaelhyi/personal-website/backend/go/util"
)

var CLOSING_H1_TAG_LENGTH = 5
var OPENING_H1_TAG_LENGTH = 4

func ValidateAndConstructPost(text *string) (Post, error) {
	var noop Post

	if util.IsStringInvalid(text) {
		return noop, errors.New("Text is invalid")
	}

	titleIndex := strings.Index(*text, "</h1>")

	if titleIndex == -1 {
		return noop, errors.New("Title cannot be blank")
	}

	newTitle := (*text)[OPENING_H1_TAG_LENGTH:titleIndex]
	newContent := (*text)[titleIndex+CLOSING_H1_TAG_LENGTH:]

	if util.IsStringInvalid(&newTitle) {
		return noop, errors.New("Title cannot be blank")
	}

	if util.IsStringInvalid(&newContent) {
		return noop, errors.New("Content cannot be blank")
	}

	newId := strings.ToLower(newTitle)
	newId = strings.ReplaceAll(
		strings.ToLower(newTitle),
		" ",
		"-",
	)

	regex := regexp.MustCompile("[^a-z-]")
	newId = regex.ReplaceAllString(newId, "")

	var id string
	if newId[len(newId)-1] == '-' {
		id = newId[:len(newId)-1]
	} else {
		id = newId
	}

	date := time.Now().Format("2006-01-02 15:04:05")
	title := strings.ReplaceAll(newTitle, "<[^>]*>", "")
	content := newContent

	post := Post{
		ID:      id,
		Date:    date,
		Title:   title,
		Content: content,
	}

	return post, nil
}

//
//    public static boolean isImageInvalid(MultipartFile image) {
//        return image == null
//                || image.isEmpty()
//                || image.getSize() == 0
//                || image.getOriginalFilename() == null
//                || image.getOriginalFilename().isEmpty()
//                || image.getOriginalFilename().isBlank()
//                || !image.getOriginalFilename()
//                        .matches(".*\\.(jpg|jpeg|png|webp)$")
//                || !image.getContentType().matches("image/(jpeg|png|webp)");
//    }
